### PR TITLE
[Transform] avoid transform failure during rolling upgrade

### LINKED ISSUE #72525: [Transform] transform fails during rolling upgrade with Failed to load transform configuration
**Affected versions:** 7.12
**Fixed in:** TBD

Transforms don't need to be stopped during a rolling upgrade, but can continue to run. A transform that runs on a to upgrade node will be automatically relocated to another node.

This relocation can fail with _Failed to load transform configuration_ if at least 2 transforms run at the time of the upgrade on the same node.

**Mitigation:**

Check the status of your transforms after the upgrade using the UI or API. In case the transform failed with the above error, `force` stop and start the transform again using the UI or API:

https://www.elastic.co/guide/en/elasticsearch/reference/current/stop-transform.html
https://www.elastic.co/guide/en/elasticsearch/reference/current/start-transform.html

As an alternative you can suspend transforms during the rolling upgrade, however this is not necessary, there are no downsides in restarting them as explained.

**Fix:**

Transform node allocation needs to check that shards are allocated.

### PR DESCRIPTION
**Pull request by hendrikmuhs**
ensure shards are searchable after creation of a new internal index version

fixes #72525

### CODE REVIEW

File: x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/persistence/TransformInternalIndexTests.java [hunk 1]
```diff
@@ -7,26 +7,37 @@
...
 import org.elasticsearch.client.AdminClient;
 import org.elasticsearch.client.Client;
+import org.elasticsearch.client.ClusterAdminClient;
 import org.elasticsearch.client.IndicesAdminClient;
 import org.elasticsearch.cluster.ClusterName;
 import org.elasticsearch.cluster.ClusterState;
 import org.elasticsearch.cluster.metadata.IndexMetadata;
 import org.elasticsearch.cluster.metadata.IndexTemplateMetadata;
 import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.metadata.Metadata.Builder;
```
Comment (przemekwitek) [inline x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/persistence/TransformInternalIndexTests.java:25]: I would remove this import and use `Metadata.Bulider` in the code.
IMO this is more readable as sole `Builder` seen in the code doesn't tell much.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformInternalIndex.java [hunk 2]
```diff
@@ -392,19 +396,43 @@ protected static boolean haveLatestVersionedIndex(ClusterState state) {
         return state.getMetadata().getIndicesLookup().containsKey(TransformInternalIndexConstants.LATEST_INDEX_VERSIONED_NAME);
     }
 
+    protected static boolean allShardsActiveForLatestVersionedIndex(ClusterState state) {
+        IndexRoutingTable indexRouting = state.routingTable().index(TransformInternalIndexConstants.LATEST_INDEX_VERSIONED_NAME);
+
+        return indexRouting != null && indexRouting.allPrimaryShardsActive();
+    }
+
     protected static boolean haveLatestAuditIndexTemplate(ClusterState state) {
```
Comment (przemekwitek) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformInternalIndex.java:405]: s/have/has/
?

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformInternalIndex.java [hunk 3]
```diff
@@ -388,23 +392,47 @@ public static void ensureLatestIndexAndTemplateInstalled(ClusterService clusterS
...
+        return indexRouting != null && indexRouting.allPrimaryShardsActive();
+    }
+
+    protected static boolean hasLatestAuditIndexTemplate(ClusterState state) {
         return state.getMetadata().getTemplates().containsKey(TransformInternalIndexConstants.AUDIT_INDEX);
     }
 
+    private static void waitForLatestVersionedIndexShardsActive(Client client, ActionListener<Void> listener) {
+        ClusterHealthRequest request = new ClusterHealthRequest(TransformInternalIndexConstants.LATEST_INDEX_VERSIONED_NAME)
+            .waitForActiveShards(ActiveShardCount.ALL);
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformInternalIndex.java:411]: We really only care if the primary shards are active right? 

I don't think this index ever really sets up replicas, but requiring tasks to wait for replicas seems like an unnecessary burden.

I wonder if waiting for a `yellow` status here is enough.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformInternalIndex.java [hunk 4]
```diff
@@ -388,23 +392,47 @@ public static void ensureLatestIndexAndTemplateInstalled(ClusterService clusterS
         );
     }
 
-    protected static boolean haveLatestVersionedIndex(ClusterState state) {
+    protected static boolean hasLatestVersionedIndex(ClusterState state) {
         return state.getMetadata().getIndicesLookup().containsKey(TransformInternalIndexConstants.LATEST_INDEX_VERSIONED_NAME);
     }
 
-    protected static boolean haveLatestAuditIndexTemplate(ClusterState state) {
+    protected static boolean allShardsActiveForLatestVersionedIndex(ClusterState state) {
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformInternalIndex.java:399]: technically, this is `allPrimaryShardsActiveForLatestVersionedIndex`

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformInternalIndex.java [hunk 5]
```diff
@@ -413,14 +441,21 @@ protected static void createLatestVersionedIndexIfRequired(
             CreateIndexRequest request = new CreateIndexRequest(TransformInternalIndexConstants.LATEST_INDEX_VERSIONED_NAME)
                 .settings(settings())
                 .mapping(mappings())
-                .origin(TRANSFORM_ORIGIN);
+                .origin(TRANSFORM_ORIGIN)
+                .waitForActiveShards(ActiveShardCount.ALL);
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/persistence/TransformInternalIndex.java:445]: I think having `ALL` here is fine as the index has our original settings, which doesn't include replicas (I think).

**Review by benwtrent — APPROVED** [review]
This change is much cleaner and nicer than the previous!

My only question is around replicas/primaries. 

On initial creation, we are guaranteed to have no replicas. So waiting for all shards there is great. But, is it possible that a user adds replicas later? This means that during a rolling upgrade, we will have to wait for replicas to get assigned in certain cases.

**Comment by hendrikmuhs** [comment]
> On initial creation, we are guaranteed to have no replicas. So waiting for all shards there is great. But, is it possible that a user adds replicas later? This means that during a rolling upgrade, we will have to wait for replicas to get assigned in certain cases.

The user isn't supposed to add replicas by himself I think, because it is a system index. The index uses `auto_expand_replicas 0-1`.

The added `waitForActiveShards` is probably not necessary:
https://github.com/elastic/elasticsearch/blob/5baabff6670a8ed49297488ca8cac8ec12a2078d/server/src/main/java/org/elasticsearch/action/admin/indices/create/TransportCreateIndexAction.java#L111

Nevertheless I thought explicit is better than implicit.

Waiting for yellow vs. green: Yes, yellow seems enough, however does that really matter? Having green sounds reasonable to me, given that the problems we are talking about are during rolling upgrade where nodes come and go.

**Comment by benwtrent** [comment]
> Yes, yellow seems enough, however does that really matter? Having green sounds reasonable to me, given that the problems we are talking about are during rolling upgrade where nodes come and go.

Green does require replicas to be assigned. So, it's waiting for a longer period of time.

But since this is a system index, a user updating the number of replicas is an unsupported case. I think the change is fine as is, just wanted to make sure we thought of this.

**Pull request merged**